### ISSUE TITLE
Audit for missing `for_each` implementations. 

### ISSUE THREAD

**Issue by jswrenn** [rust-itertools/itertools#755]
Implementing `for_each` for our iterators allows us to provide more performant internal iteration. We provide it in some cases, but not all. I'd like us to take three steps towards resolving this:
1. Take an inventory of whether our adapters provide `for_each`.
2. Implement `for_each` for missing adapters. 
3. (Potentially!) Implement a clippy lint to prevent regressions.

Happy to provide mentorship for any of these items.

**Comment by scottmcm** [comment]
Note that the vast majority of the time, you should be overriding `fold`, not `for_each`.

The default `for_each` just calls `fold` with a trivial `()` accumulator, which is easily optimized out (since in the default Rust calling convention, it's not even passed).

Thus the only reason to override `for_each` *in addition* to `fold` is if you can optimize something by knowing that the accumulation is stateless, but I can't off-hand recall any case where I've seen that to be the case.

**Comment by jswrenn** [comment]
Whoops, yes, of course. I'll modify the issue accordingly.

**Comment by Philippe-Cholet** [comment]
First, find `fold` implementations is easy.

file | Adaptors with specialized `fold`
--- | -----------------------------
adaptors/coalesce.rs | CoalesceBy
adaptors/map.rs | MapSpecialCase
adaptors/mod.rs | PutBack, Product, FilterOk, FilterMapOk, Update
cons_tuples_impl.rs | ConsTuples
intersperse.rs | IntersperseWith
process_results_impl.rs | ProcessResults

Other adaptors and other iterators do not have `fold` specializations.

**Comment by phimuemue** [comment]
Once we do this, we should test all of them via `test_specializations`.

**Comment by Owen-CH-Leung** [comment]
I'm interested to contribute to this crate and I digged a bit into the source code. I think one potential candidate to override `fold` is `Combinations`. What do you guys think ?

I can create a PR in coming few days to implement this.

**Comment by phimuemue** [comment]
> I'm interested to contribute to this crate and I digged a bit into the source code. I think one potential candidate to override `fold` is `Combinations`.
> 
> https://github.com/rust-itertools/itertools/blob/master/src/combinations.rs#L100
> 
> What do you guys think ?
> 
> I can create a PR in coming few days to implement this.

@Philippe-Cholet I think you're the most knowledgeable there. Could you mentor/review this?

**Comment by Philippe-Cholet** [comment]
I previously had some thoughts about spliting `next` in two methods to make a `nth` that would create only one vector created instead of `n+1`. And I earlier thought it would be better done before `fold` but no as it needs to create each item anyway. I guess `fold` would be an interesting and total rewrite of `next`.
@Owen-CH-Leung Hi, sure you can help. And I'll be there to "mentor/review" it.

**Comment by DavidCoy77** [comment]
@jswrenn, I'd love to get started contributing to Rust, but I'd need a little help getting started since this would be my first ever GitHub contribution. You mention you're happy to provide mentorship in your initial post, so I'm eager to get to work if you could help me out with any of the items on the checklist. Thanks!

**Comment by Philippe-Cholet** [comment]
@DavidCoy77 I guess that specialize `PutBackN::fold` would be a good first pull request. You could look at #772 to find out what's expected.

**Comment by jswrenn** [comment]
Hi @DavidCoy77, apologies for not replying to your emails — I've been traveling. I agree with @Philippe-Cholet that our fold implementations are the best place to start.

**Comment by DavidCoy77** [comment]
Great! Thank you for the help @jswrenn and @Philippe-Cholet. I've spent some time looking at the `WithPosition::fold` at #772 Philippe mentioned and unfortunately I have a hard time what would be asked of me for `PutBackN::fold` , so If you could offer additional guidance it would be greatly appreciated.

**Comment by Philippe-Cholet** [comment]
I guess there is a lot of new things (I did not know criterion and quickcheck before coming here, and was not that familiar with `fold` either).

- Are you familiar with `Iterator::fold`? With it, we consume the iterator and therefore consume all items. Specialize it can be faster than calling `next` repeatedly as it would be the case without specialization. You might not use `fold` directly but by default there are a few iterator methods that use `fold` such as `for_each` which I believe is more well known/used(?).
- Are you familiar with `itertools::PutBackN`? I know I was not but I like the doctest example.
- The core of the pull request would be to write `PutBackN::fold`.
- In "benches/bench1.rs", we would write a benchmark function (using criterion) to measure the time improvement of specializing `PutBackN::fold` as we don't want to slow things down. Soon enough, this should be automatically provided by a macro I'm writing in another PR.
- In "tests/specializations.rs", we write a function inside a "quickcheck" macro (it's property based testing, basically it tests our function against many inputs, and when it fails, it tries to find the smallest input that fails, super useful to find edge cases we did not consider properly). In that file, `test_specializations` takes a clonable iterator and test "fold" specialization amongst other method specializations, which simply allow us to write less code.

You can focus on figuring out `PutBackN::fold` first.

    pub struct PutBackN<I: Iterator> {
        top: Vec<I::Item>,
        iter: I,
    }
    // From the next method, we want elements of `top` in the reversed order then, only then, the elements from `iter` (normal order).
    fn next(&mut self) -> Option<Self::Item> {
        self.top.pop().or_else(|| self.iter.next())
    }

**Comment by Philippe-Cholet** [comment]
@jswrenn I'm curious but don't understand the exact intention behind this yet:

> 3. (Potentially!) Implement a clippy lint to prevent regressions.

A lint that warn us about missing `fold` specializations, is that it?!

**Comment by jswrenn** [comment]
Yes, the idea is to add lints to clippy itself to warn us about missing `fold` specializations. For any adaptors where we explicitly didn't want to specialize `fold`, we would need to write `#[allow(clippy::missing_iterator_fold)]`. This would allow us to audit for missing clippy lints as simple as running `cargo clippy`, and to provide an automated warning if we ever tried to add a new adaptor without a fold implementation.

**Comment by luander** [comment]
I'm looking to give this issue a try. 
Started with [Combinations](https://github.com/rust-itertools/itertools/blob/master/src/combinations.rs) and I wonder how does folding on that one would look like? with `LazyBuffer` and such.
I can also use some mentoring/walkthrough of the crate if someone is willing to show me around.

**Comment by Philippe-Cholet** [comment]
@luander I can happily mentor you on this.
- Specializations are tested in "tests/specializations.rs" so test it is a command line away (`cargo test combinations` is fast enough).
- Specializations are benchmarked in "benches/specializations.rs" so benchmark it would be a command line away too (`cargo bench --bench specializations combinations/fold` to ignore others). But build this particular benchmark is really slow (almost 2 minutes for me) so the simplest thing to do is to temporarily comment out unrelated benchmarks (without commiting).
- I'm not sure what it would look like exactly. I guess we might want to do some work from `next` while the lazy buffer is not full yet then some simpler work once we know for sure it is full.
  In the process, we might want to expand `LazyBuffer` abilities.
  Specialize `Combinations::fold` would definitely not be as straightforward as other iterators but it's an interesting one.

**Comment by phimuemue** [comment]
> @jswrenn I'm curious but don't understand the exact intention behind this yet:
> 
> > 3. (Potentially!) Implement a clippy lint to prevent regressions.
> 
> A lint that warn us about missing `fold` specializations, is that it?!

If clippy does not accept the lint, we could maybe (cleverly ab)use [`missing_trait_methods`](https://rust-lang.github.io/rust-clippy/master/#/missing_trait_methods).

**Comment by Philippe-Cholet** [comment]
**Update on the lint idea:** After discussion on zulip, [`missing_trait_methods`](https://rust-lang.github.io/rust-clippy/master/#/missing_trait_methods) _might_ accept a configuration (in `clippy.toml`) instead of a new lint that would be too specific to a single trait method when it could be generalized. It would not be as flexible as we might want but it would be decent.
There are alternatives though: fork clippy (but maintainance?), use clippy sister project [marker](https://github.com/rust-marker/marker/) (great developper experience but it's a work in progress, I can't identify an iterator implementation with it yet), [dylint](https://crates.io/crates/cargo-dylint) (should be nice enough but my first try was a failure).

**Comment by Philippe-Cholet** [comment]
@phimuemue Oh... I find embarrassing I was aware about `--message-format json` for "check" (never used) but not for "clippy".

After some tinkering on [jq playground](https://jqplay.org/), I think I have a new git alias (it's not git-related but useful anyway):

	missing-trait-methods = "!f() { cargo clippy --quiet --message-format json -- --warn clippy::missing_trait_methods | jq '. | select(.message.code.code == \"clippy::missing_trait_methods\") | select(.message.message | endswith(\"`'$2'`\")) | .message.spans[0] | .file_name + \":\" + (.line_start | tostring) + \"    \" + .text[0].text' | grep $1; }; f"

then

    git missing-trait-methods Iterator fold
    git missing-trait-methods DoubleEndedIterator rfold
<details><summary>Output for `git missing-trait-methods Iterator size_hint`</summary>
<p>

```
"src\\adaptors\\mod.rs:498    impl<B, F, I> Iterator for Batching<I, F>"
"src\\groupbylazy.rs:387    impl<'a, K, I, F> Iterator for Groups<'a, K, I, F>"
"src\\groupbylazy.rs:440    impl<'a, K, I, F> Iterator for Group<'a, K, I, F>"
"src\\groupbylazy.rs:556    impl<'a, I> Iterator for Chunks<'a, I>"
"src\\groupbylazy.rs:599    impl<'a, I> Iterator for Chunk<'a, I>"
"src\\grouping_map.rs:25    impl<K, V, I, F> Iterator for MapForGrouping<I, F>"
"src\\sources.rs:127    impl<A, St, F> Iterator for Unfold<St, F>"
```
</p>
</details>

**Comment by Philippe-Cholet** [comment]
@luander Can I help you more on `Combinations::fold`?
Alternatively, I suggested `ZipEq::fold` but if you want to work on `Combinations`, specialize `nth` would be appreciated (it would solve #301 and close the outdated and stuck #329).

**Comment by Philippe-Cholet** [comment]
@jswrenn A lint `missing_iterator_fold` was not particularly well received: too specific to iterators basically.
The main alternative: `clippy::missing_trait_methods` could be configurable to limit warnings (6538 right now):

    # clippy.toml
    missing_trait_methods = [
        "core::iter::Iterator::fold",
        "core::iter::DoubleEndedIterator::rfold",
        "core::iter::Iterator::size_hint",
    ]

The downside is that allow the lint on a `impl Iterator` would allow not only one missing specialization but all of them. But if we limit ourselves to `size_hint/fold/rfold` then this seems good enough to me.

_Is that okay with you?_ If it is, I'll work on making the lint configurable.

**Comment by kinto-b** [comment]
@Philippe-Cholet Hi there, on `Combinations::fold`, once #914 is merged, I think something like this should work as the core of an implementation:

```rust
// increment_indices returns true once we run out of combinations
while !increment_indices() {
  combination = self.indices.iter().map(|i| self.pool[*i].clone()).collect();
  acc = f(acc, combination)
}
```

If you think that looks reasonable I'll put through a PR.

**Comment by Philippe-Cholet** [comment]
> I tried this out and, although it works, it is markedly slower than the unspecialized implementation

@kinto-b We won't specialize it if it's slower (but that alone is a valuable information). About combinations, the slowest thing is heap-allocate the generated items. `fold` by nature generates them all (to give them to `f`) so we can't win anything on that side.
The only thing `Combinations::fold` could do is simpler code after the initialization phase while the default `fold` repeatedly doing `next` have "init code" lying around without use after a few steps, basically what you did but even it was a perf win, I did not expect much out of it since the heap-allocation is the bottleneck here and nothing else really impact performance (or at least I think so).
I kinda wonder if mutate a combination IN-PLACE and give clones would be faster but _it's a lot of work for a probable small win (if any)_: `while !self.increment_indices_and_comb(&mut comb) { acc = f(acc, comb.clone()); }` or something like that.

I will eventually work again on configure `clippy::missing_trait_methods` and (if accepted) we will be able to mark that some `fold` are allowed to not be specialized and ensures all other `fold` are.

**Comment by khoover** [comment]
Two things: short first one is this should be recreated for `try_fold` as well since that's stable. EDIT: Ah, that might need `Try` trait stabilization; maybe behind a `nightly` feature?

Second, re: combinations, it seems like you have two separate cases. `ArrayCombinations` could be reasonably implemented, since there's no allocation beyond the stack space for the array. But for the base `Combinations` case, maybe instead of implementing fold there's a
```rust
pub fn fold_over_iter<S, F>(self, init: S, f: F) -> S
where
  F: for<'a> FnMut(S, FoldCombinationsIter<'a, I>) -> S
```
since I'd believe most uses of such a fold wouldn't actually care whether the combinations are given as a Vec or an iterator.

**Comment by repairman29** [comment]
🎯 **PING RECEIVED!**

**@repairman29** has claimed this bounty via [Echeo](https://echeo.io).

**Match Score:** 85%  
**Status:** Ready to start working!

[View conversation in Echeo](https://echeo.io/messages) to coordinate.

**Comment by repairman29** [comment]
Hi! I'm interested in contributing to this. Let me know if you'd like to discuss the implementation approach.

*(Connected through [Echeo](https://echeo.io))*
